### PR TITLE
DPC++ implementation of GPU backend for K-Means initialization

### PR DESCRIPTION
**Pull request by amgrigoriev**

### CODE REVIEW

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#include "oneapi/dal/algo/kmeans_init/common.hpp"
+
+namespace oneapi::dal::kmeans_init::backend {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+namespace pr = dal::backend::primitives;
+
+template <typename T>
+struct copy_first_observations {};
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp:32]: Please delete this

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        const std::int64_t column_count = centroids.get_dimension(1);
+        const auto data_ptr = data.get_data();
+        auto centroids_ptr = centroids.get_mutable_data();
+        auto copy_event = queue.submit([&](sycl::handler& cgh) {
+            cgh.parallel_for<copy_first_observations<Float>>(
+                sycl::range<1>(column_count * cluster_count),
+                [=](sycl::id<1> idx) {
+                    centroids_ptr[idx] = data_ptr[idx];
+                });
+        });
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp:70]: Please use `q.copy` instead

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp (hunk 1 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp]: Done

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp (hunk 2 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp]: Done

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp [hunk 3]
```diff
@@ -30,13 +26,8 @@ namespace oneapi::dal::kmeans_init::backend {
 using std::int64_t;
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp]: Done

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp [hunk 4]
```diff
@@ -30,13 +26,8 @@ namespace oneapi::dal::kmeans_init::backend {
...
 namespace interop = dal::backend::interop;
-
-template <typename Float, typename Method>
-using daal_kmeans_init_kernel_t =
-    daal_kmeans_init::internal::KMeansInitDenseBatchKernelUCAPI<to_daal_method<Method>::value,
-                                                                Float>;
+namespace pr = dal::backend::primitives;
 
 template <typename Float, typename Method, typename Task>
 static compute_result<Task> call_daal_kernel(const context_gpu& ctx,
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp]: Done

File: cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
+/*******************************************************************************
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp]: Totally reworked

File: cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+#include <daal/src/algorithms/engines/engine_batch_impl.h>
+
+#include "oneapi/dal/array.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+template <typename T>
+class rnd_uniform {
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp]: Done

File: cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        static_assert(std::is_integral_v<T> || std::is_floating_point_v<T>);
+        ONEDAL_ASSERT(*engine_ != nullptr);
+        auto engine_impl =
+            dynamic_cast<daal::algorithms::engines::internal::BatchBaseImpl*>(&(*engine_));
+
+        auto* values = result_array.get_mutable_data();
+        auto count = result_array.get_count();
+        const size_t count_as_size_t = dal::detail::integral_cast<std::size_t>(count);
+
+        daal::internal::RNGs<size_t, daal::sse2> rng;
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp]: Done

File: cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+template <typename T>
+class rnd_uniform {
+public:
+    rnd_uniform() : engine_(daal::algorithms::engines::mt19937::Batch<>::create(777)) {}
+    void generate(sycl::queue& queue, ndview<T, 1>& result_array, T a, T b) {
+        static_assert(std::is_integral_v<T> || std::is_floating_point_v<T>);
+        ONEDAL_ASSERT(*engine_ != nullptr);
+        auto engine_impl =
+            dynamic_cast<daal::algorithms::engines::internal::BatchBaseImpl*>(&(*engine_));
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp]: Done

File: cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+template <typename T>
+class rnd_uniform {
+public:
+    rnd_uniform() : engine_(daal::algorithms::engines::mt19937::Batch<>::create(777)) {}
+    void generate(sycl::queue& queue, ndview<T, 1>& result_array, T a, T b) {
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp]: Fixed

File: cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        ONEDAL_ASSERT(*engine_ != nullptr);
+        auto engine_impl =
+            dynamic_cast<daal::algorithms::engines::internal::BatchBaseImpl*>(&(*engine_));
+
+        auto* values = result_array.get_mutable_data();
+        auto count = result_array.get_count();
+        const size_t count_as_size_t = dal::detail::integral_cast<std::size_t>(count);
+
+        daal::internal::RNGs<size_t, daal::sse2> rng;
+        rng.uniform(count_as_size_t, values, engine_impl->getState(), a, b);
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp]: Reworked

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+            dal::detail::integral_cast<std::uint64_t>(sizeof(std::int32_t)));
+
+        auto indices = pr::ndarray<std::uint64_t, 1>::empty(queue, cluster_count);
+        auto indices_ptr = indices.get_mutable_data();
+
+        pr::rnd_uniform<std::uint64_t> rng;
+        bk::event_vector events;
+
+        std::int64_t k = 0;
+        for (std::uint64_t i = 0; i < cluster_count; i++) {
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp]: Done

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            rng.generate(queue, cur_elements, i, row_count);
+            ONEDAL_ASSERT(indices_ptr[i] >= 0)
+            std::uint64_t& value = indices_ptr[i];
+            for (std::uint64_t j = i; j > 0; j--) {
+                if (value == indices_ptr[j - 1]) {
+                    value = j - 1;
+                }
+            }
+            if (value >= row_count)
+                continue;
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp]: Moved to primitive as a partial shuffle method

File: cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    rnd_uniform() : engine_(daal::algorithms::engines::mt19937::Batch<>::create(777)) {}
+    void generate(sycl::queue& queue, ndview<T, 1>& result_array, T a, T b) {
+        static_assert(std::is_integral_v<T> || std::is_floating_point_v<T>);
+        ONEDAL_ASSERT(*engine_ != nullptr);
+        auto engine_impl =
+            dynamic_cast<daal::algorithms::engines::internal::BatchBaseImpl*>(&(*engine_));
+
+        auto* values = result_array.get_mutable_data();
+        auto count = result_array.get_count();
+        const size_t count_as_size_t = dal::detail::integral_cast<std::size_t>(count);
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/rng/rnd_uniform.hpp]: Done

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp [hunk 14]
```diff
@@ -14,91 +14,67 @@
...
 
-using std::int64_t;
 using dal::backend::context_gpu;
 
-namespace daal_kmeans_init = daal::algorithms::kmeans::init;
 namespace interop = dal::backend::interop;
+namespace pr = dal::backend::primitives;
+
+template <typename Task>
+using result_t = compute_result<Task>;
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp:32]: Looks like we will have only one task for foreseeable future, so the `Task` can be substituted with concrete type. The same is valid for the rest of the kernel.

File: cpp/oneapi/dal/backend/primitives/rng/partial_shuffle.hpp [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+/*Partial Fisher-Yates Shuffle algorithm*/
+struct partial_shuffle {
+    void generate(ndview<size_t, 1>& result_array, size_t top);
+};
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/rng/partial_shuffle.hpp:28]: Does it make sense to have structure here? Would the function be enough?

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp (hunk 14 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp]: Fixed

File: cpp/oneapi/dal/backend/primitives/rng/partial_shuffle.hpp (hunk 15 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/rng/partial_shuffle.hpp]: Changed to function

File: cpp/oneapi/dal/backend/primitives/rng/partial_shuffle.cpp [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+template <typename... Args>
+inline void uniform_by_cpu(Args&&... args) {
+    dispatch_by_cpu(context_cpu{}, [&](auto cpu) {
+        daal::internal::
+            RNGs<size_t, oneapi::dal::backend::interop::to_daal_cpu_type<decltype(cpu)>::value>{}
+                .uniform(std::forward<Args>(args)...);
+    });
+}
+
+void partial_fisher_yates_shuffle(ndview<size_t, 1>& result_array, size_t top) {
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/rng/partial_shuffle.cpp:45]: Done

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp [hunk 17]
```diff
@@ -14,93 +14,65 @@
...
 
-namespace daal_kmeans_init = daal::algorithms::kmeans::init;
 namespace interop = dal::backend::interop;
+namespace pr = dal::backend::primitives;
+
+using result_t = compute_result<task::init>;
+
+using descriptor_t = detail::descriptor_base<task::init>;
+
+using input_t = compute_input<task::init>;
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp:35]: Remove extra spaces

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp [hunk 18]
```diff
@@ -14,93 +14,65 @@
...
     const int64_t column_count = data.get_column_count();
+    const int64_t row_count = data.get_row_count();
     const int64_t cluster_count = params.get_cluster_count();
 
-    daal_kmeans_init::Parameter par(dal::detail::integral_cast<std::size_t>(cluster_count));
-
-    const auto daal_data = interop::convert_to_daal_table(queue, data);
+    auto data_ptr =
+        row_accessor<const Float>(data).pull(queue, { 0, -1 }, sycl::usm::alloc::device);
+    auto arr_data = pr::ndarray<Float, 2>::wrap(data_ptr, { row_count, column_count });
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp:50]: We have function called `table2ndarray` in `primitives/utils.hpp`

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+
+template <typename Float>
+struct kmeans_init_kernel<Float, kmeans_init::method::random_dense> {
+    static sycl::event compute_initial_centroids(sycl::queue& queue,
+                                                 const pr::ndview<Float, 2>& data,
+                                                 pr::ndview<Float, 2>& centroids) {
+        ONEDAL_ASSERT(data.get_dimension(1) == centroids.get_dimension(1));
+        ONEDAL_ASSERT(data.get_dimension(0) >= centroids.get_dimension(0));
+        const std::uint64_t row_count =
+            dal::detail::integral_cast<std::uint64_t>(data.get_dimension(0));
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp:75]: Why do we need to cast to `uint`?

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+
+        auto indices = pr::ndarray<size_t, 1>::empty(queue, cluster_count);
+        partial_fisher_yates_shuffle(indices, row_count);
+        auto indices_ptr = indices.get_data();
+
+        const std::int64_t required_local_size = bk::device_max_wg_size(queue);
+        const std::int64_t local_size = std::min(bk::down_pow2(column_count), required_local_size);
+
+        auto gather_event = queue.submit([&](sycl::handler& cgh) {
+            sycl::stream out(1024, 256, cgh);
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp:93]: Remove?

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl_dbl_dpc.cpp [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+#include "oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp"
+
+namespace oneapi::dal::kmeans_init::backend {
+
+#ifdef ONEDAL_DATA_PARALLEL
+template struct kmeans_init_kernel<double, kmeans_init::method::dense>;
+template struct kmeans_init_kernel<double, kmeans_init::method::random_dense>;
+template struct kmeans_init_kernel<double, kmeans_init::method::plus_plus_dense>;
+template struct kmeans_init_kernel<double, kmeans_init::method::parallel_plus_dense>;
+#endif
```
Comment (rlnx) [inline cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl_dbl_dpc.cpp:26]: There is no need for separate float/double instantiation any more, the problem is solved now

File: cpp/oneapi/dal/backend/primitives/rng/partial_shuffle.cpp [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+    if (engine.get() == nullptr) {
+        throw internal_error(msg::failed_to_generate_random_numbers());
+    }
+    auto engine_impl =
+        dynamic_cast<daal::algorithms::engines::internal::BatchBaseImpl*>(&(*engine));
+    if (engine_impl == nullptr) {
+        throw internal_error(msg::failed_to_generate_random_numbers());
+    }
+    const std::uint64_t count = result_array.get_count();
+    const auto casted_count = dal::detail::integral_cast<size_t>(count);
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/rng/partial_shuffle.cpp:58]:
```suggestion
    const auto casted_count = dal::detail::integral_cast<std::size_t>(count);
```

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp (hunk 20 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp]: Removed. And fixed a hidden bug in data transfer.

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp (hunk 18 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp]: Replaced

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp (hunk 17 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc.cpp]: done

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp (hunk 19 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernels_impl.hpp]: removed casting

File: cpp/oneapi/dal/backend/primitives/rng/partial_shuffle.cpp [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+    if (engine.get() == nullptr) {
+        throw internal_error(msg::failed_to_generate_random_numbers());
+    }
+    auto engine_impl =
+        dynamic_cast<daal::algorithms::engines::internal::BatchBaseImpl*>(&(*engine));
+    if (engine_impl == nullptr) {
+        throw internal_error(msg::failed_to_generate_random_numbers());
+    }
+    const auto casted_top = dal::detail::integral_cast<std::size_t>(top);
+    const std::uint64_t count = result_array.get_count();
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/rng/partial_shuffle.cpp:56]: Why this is `uint64_t`?

Comment (amgrigoriev) [reply cpp/oneapi/dal/backend/primitives/rng/partial_shuffle.cpp]: Removed

**Review by rlnx — APPROVED** [review]

**Pull request merged**